### ISSUE TITLE
Recognize `backoff` in responses

### ISSUE THREAD

**Issue by vermiculus** [vermiculus/sx.el#188]
More info in #187.

**Comment by Bruce-Connor** [comment]
I agree. sx-request can check the value of backoff, add that to current time, save to a variable, and sleep before making a request if the variable is greater than current time. 

Blocking the interface is slightly annoying, but it's better than throwing an error I think. To the user, the interface already blocks momentarily while making requests anyway, so it probably won't be noticeable.

**Comment by vermiculus** [comment]
I wanted to avoid making a blocking call like that since I don't know how long a `backoff` could be – if it's something like 30 seconds or so, then it becomes a problem :)

Here's my thought process:

1. Check `sx-request-backoff-until`
 - if the clock is past that time, continue with request.
 - if that time has not yet come, signal an error
1. When we get a `backoff` request: `current-time + backoff -> sx-request-backoff-until`

**Comment by Bruce-Connor** [comment]
Well, you said the bot succeeded at 4 requests per second, so I thought the backoff would be very short. 

Blocking for several seconds is definitely unacceptable, but I don't like the idea of signaling an error just because we didn't want to wait 0.5 seconds or something. 

I see two possibilities (after performing the logic you described above, but before throwing an error):

- Have a sleep threshold of 0.5 sec or something. When we're about to make a request, if current time is below the `backoff-until` variable by less than the threshold, we sleep. Otherwise we throw an error. 

- shamelessly sleep and block the interface whenever necessary (as detected by `backoff-until`). But, immediately before sleeping, print a message like *"Waiting for X seconds, hit C-g to cancel."*

I kinda like the second approach. It lets the user decide how long he is willing to wait, and it feels emacsy.

**Comment by vermiculus** [comment]
I *like* that second idea! I agree – it feels very Emacsy. I think this is the approach we should take. I'll work on implementing this tomorrow.
